lib/pulse_sensor.c

```diff
@@ -0,0 +1,332 @@
+int setup_pulse_sensor_vss(
+		TIM_HandleTypeDef* htim,
+		U32 channel,
+		float conversionRatio,
+		float* resultStoreLocation,
+		U16 dmaStoppedTimeoutMS,
+		bool useVariableSpeedSampling,
+		U16 lowPulsesPerSecond,
+		U16 highPulsesPerSecond,
+		U16 minSamples
+		)
+{
+	PulseSensor* newSensor = &pulseSensor[numSensors];
+
+	// Set the local values of the pulse sensor data struct to the given parameters
+	newSensor->htim = htim;
+	newSensor->channel = channel;
+	newSensor->conversionRatio = conversionRatio;
+	newSensor->resultStoreLocation = resultStoreLocation;
+	newSensor->useVariableSpeedSampling = useVariableSpeedSampling;
+	newSensor->dmaStoppedTimeoutMS = dmaStoppedTimeoutMS;
+	newSensor->lowPulsesPerSecond = lowPulsesPerSecond;
+	newSensor->highPulsesPerSecond = highPulsesPerSecond;
+	newSensor->minSamples = minSamples;
+
+	U32 clockFrequency = HAL_RCC_GetSysClockFreq() * 0.5; // Multiply by 0.5 as timers are 1/2 the clock frequency
+
+	// Evaluate the size of the given timer by checking if the Auto reload value is the max size of a 16bit value 0xFFFF
+	// TODO: Change clock frequency if certain timers
+	if (htim->Instance->ARR == 0xFFFF) {
+		newSensor->timerSize = 16;
+	} else if (htim->Instance->ARR == 0xFFFFFFFF) {
+		newSensor->timerSize = 32;
+	} else {
+		newSensor->timerSize = 32;
+		return -1;
+	}
+
+	newSensor->timerPeriodSeconds = 1 / ((float)clockFrequency / (htim->Instance->PSC + 1));
+
+	// Clear the sensor's buffer
+	memset(newSensor->buffer, 0, sizeof(U32)*IC_BUF_SIZE);
+
+	// Default other values to 0
+	newSensor->lastDMAReadValueTimeMs = 0;
+	newSensor->DMALastReadValue = 0;
+	newSensor->stopped = true;
+	newSensor->vssSlope = (MAX_DELTAS - minSamples) / (float)(highPulsesPerSecond - lowPulsesPerSecond); // Calculate the constant slope value for vss once at the beginning here.
+
+	HAL_TIM_IC_Start_DMA(htim, channel, (U32*)pulseSensor[numSensors].buffer, IC_BUF_SIZE);
+
+	numSensors++;
+
+	return 0;
+}
+
+// Function for setting up timer without variable speed sampling (vss)
+int setup_pulse_sensor(
+		TIM_HandleTypeDef* htim,
+		U32 channel,
+		float conversionRatio,
+		float* resultStoreLocation,
+		U16 dmaStoppedTimeoutMS
+		)
+{
+	return setup_pulse_sensor_vss(
+		htim,
+		channel,
+		conversionRatio,
+		resultStoreLocation,
+		dmaStoppedTimeoutMS,
+		false,
+		0,	// Low RPM Value
+		0,	// High RPM Value
+		0 	// Min samples
+		);
+}
+
+// Function for going through all of the currently set up pulse sensors
+int check_pulse_sensors() {
+	for (int sensorNumber = 0; sensorNumber < numSensors; sensorNumber++) {
+		evaluate_pulse_sensor(sensorNumber);
+	}
+
+	return 0;
+}
+
+// Function that goes through the buffer of the given pulse sensor, handling edge cases, and setting the return value to the found speed value.
+int evaluate_pulse_sensor(int sensorNumber) {
+
+	// Get the current tick and use it throughout the whole function so it can't change between operations.
+	U32 currentTick = HAL_GetTick();
```

Comment:
Just a heads up, I have been burned by using the currentTick trick before, as during the execution of the code the tick will change, messing up the values. Change to just using HAL_GetTick() everywhere

Reply:
I used to have everything do that, but I was concerned about that tick changing and potentially having a tiny chance of entering a stopped state when it isn't supposed to because a tick changed at just the right moment. Since the values I'm using where this could happen aren't changing even though the tick will be when I look at them, wouldn't it be better to store it here and have less moving parts?

Reply:
That makes sense, keep doing it this way